### ISSUE TITLE
Race condition with <script type=systemjs-module>

### ISSUE THREAD

**Issue by joeldenning** [systemjs/systemjs#2040]
See [codepen for demo of this](https://codepen.io/joeldenning/project/editor/Aapxan#0)

Under the following conditions, systemjs throws an error:

- Using named-register extra
- `<script type="systemjs-module" src="./index.js">` in document **before system.js is loaded**
- The module loaded via `type=systemjs-module` has a dependency on another module.

If the module depended on is in the import map, you get this error:
<img width="435" alt="image" src="https://user-images.githubusercontent.com/5524384/66249381-a6b59100-e6ef-11e9-8f0a-34afacaf5694.png">

If the module depended on is a url instead of bare module specifier, you get this error:
<img width="429" alt="image" src="https://user-images.githubusercontent.com/5524384/66249389-d4023f00-e6ef-11e9-8126-fa4aebc95031.png">

**Comment by guybedford** [comment]
Unfortunately there is no guarantee that the timeout will ensure enough time for the extras to load.

The full-proof fix would be for every extra to hook a SystemJS call that re-checks for systemjs-module scripts.

I don't like it though at all to be honest as the sensible fix would be to:

1. Build SystemJS and the extras into a single script
2. Load modules after SystemJS itself is loaded.

The real root cause here is that systemjs-module before systemJS gives the best load timings.

But systemjs-module after SystemJS is really what makes sense.

The mismatch is that we're unable to load systemjs-module after SystemJS quickly enough.

If anything this makes me think systemjs-module is bad and we should remove it and instruct users to use `<script>System.import</script>`.

I'm really not sure what benefits systemjs-module is offering given these costs.

**Comment by joeldenning** [comment]
> I'm really not sure what benefits systemjs-module is offering given these costs.

I find value in systemjs-module because it brings support for the `import:` syntax and also is really close to how modules are loaded natively. Improves systemjs' polyfilling story. Additionally, it is best load timing / performance to use `<script type="systemjs-module">` before systemjs is loaded.

I think the bug here can be solved a different way. The problem is that the named-register extra creates an entirely separate instance of the SystemJS loader, but any system imports that happen before that will have the old loader in their closure and so async calls to instantiate/resolve/etc will use the old loader instance instead of the new one.

Wouldn't just not creating a new instance of the SystemJS loader be the easiest fix for this? We can add [registerRegistry](https://github.com/systemjs/systemjs/blob/master/src/extras/named-register.js#L14) to the existing loader, while also patching the constructor. I think that would solve this?

https://github.com/systemjs/systemjs/blob/master/src/extras/named-register.js

**Comment by joeldenning** [comment]
I created #2042 which addresses this. However, it doesn't solve the deeper underlying issue of System.import() being called before the extras are loaded.


> Unfortunately there is no guarantee that the timeout will ensure enough time for the extras to load.

Are you sure about this? Before DOMContentLoaded when the browser is executing the html file sequentially, does it allow for any ticks of the event loop in between `<script>`s?? Do you have proof of this? A link? Codepen?

I'm trying to brainstorm other problems. I like the idea of a `system.full.js` bundle that just has everything in it.

Options:
1. `system.full.js` with everything in it. This would also be easier for people to use just in general because many people get confused by which extras they need and which order to load them in.
2. A new `module-scripts.js` extra that is placed after all other extras and triggers the System.import for all the `<script type=systemjs-module>` scripts. This would mean systemjs@7 since the functionality would be removed from script-load.js
3. Merge in #2042 which solves one problem with things but does solve the one problem we're aware of.

**Comment by joeldenning** [comment]
> Unfortunately there is no guarantee that the timeout will ensure enough time for the extras to load.

Okay actually I just confirmed that you are correct on this, by testing a small example out locally. It's network-dependent.

**Comment by joeldenning** [comment]
I think I lean towards Option 4 -- it's the one that doesn't require a breaking change and is covers all use cases. It sacrifices performance, but is an opt-in behavior. Those with stricter performance needs can just not use the feature.

**Comment by guybedford** [comment]
I like (2) the most.

(4) isn't the ideal performance we want since script fetching must always be as eager as possible on the network. A TCP head of line block of the main HTML content should not result in a delayed fetch request for the scripts.

**Comment by joeldenning** [comment]
Reopening because #2042 doesn't fully fix this.

@guybedford you lean towards a new module-scripts extra?

**Comment by guybedford** [comment]
To be honest I don't actually think there is anything to be done here. This is just is the way things have to be to be honest. If you want exact timing use `System.import`. If you need it to load after extras place systemjs-module after SystemJS itself and it will run on contentloaded. It's use-case specific.

**Comment by joeldenning** [comment]
So you're leaning towards just chalking it up as a limitation of the implementation, maybe making a note of it in the docs?

**Comment by guybedford** [comment]
Yes it's important for users to understand load timings.

Personally I would -

1. Revert to documenting `<script>System.import()</script>` as the main import method, removing systemjs-module examples
2. When mentioning systemjs-module, include `<script type="systemjs-module">` before SystemJS itself, and make a note that loading will start before extras unless extras are bundled.

**Comment by guybedford** [comment]
I think if anything we should treat this as a documentation problem at this point. @joeldenning are there any further docs changes you think we could add here to resolve this?

**Comment by joeldenning** [comment]
Perhaps we should update the systemjs-examples repo, too?

Specifically:
- https://github.com/systemjs/systemjs-examples/blob/7ea5cdc139e5210713d0e646983dd312317bb548/systemjs-features/basic-import-map/index.html#L17
- https://github.com/systemjs/systemjs-examples/blob/7ea5cdc139e5210713d0e646983dd312317bb548/loading-code/react-hello-world/index.html#L20
- https://github.com/systemjs/systemjs-examples/blob/7ea5cdc139e5210713d0e646983dd312317bb548/loading-code/react-hello-world/README.md#L7

**Comment by guybedford** [comment]
That would probably make sense, yes.

**Comment by joeldenning** [comment]
@guybedford I've come around to Option 2 in https://github.com/systemjs/systemjs/issues/2040#issuecomment-538617450. I think it might be time to discuss systemjs@7, with removed use-default extra, moving systemjs-module support out to an extra, css modules, multiple import maps support, and other things mentioned in https://github.com/systemjs/systemjs/issues/2095.

**Comment by guybedford** [comment]
@joeldenning I'd be happy to move the script support out to an extra, but I don't see any rush to release this do you? Perhaps we can focus on the docs and examples first?

As for use-default, again I'm not sure it's doing damage sitting there as long as docs include necessary info?

CSS modules etc are an extra, so again experimental status seems ok to me. I would rather wait until further spec progress here before doing anything.

For multiple import maps as well, this is a useful feature that is still awaiting spec progress, so as long as we are clear it is experimental that also seems fine.

I think major releases should be focused on when we know the directions to push for, not when we don't.

**Comment by joeldenning** [comment]
> I think major releases should be focused on when we know the directions to push for, not when we don't.

Fair enough, I agree.

**Comment by TheMcMurder** [comment]
Just to verify I understand this limitation while we're waiting on the documentation changes. 

- importing with `system.import` waits for all extras to be finished first
- importing with `<script type=systemjs-module>` does not

**Comment by joeldenning** [comment]
Very close:

- importing with system.import waits for all extras to be finished first
- `<script type=systemjs-module>` **above the script that loads systemjs** does not wait for extras.
- `<script type=systemjs-module>` **below the script that loads systemjs** does wait for extras (via [this DOMContentLoaded listener](https://github.com/systemjs/systemjs/blob/master/src/features/script-load.js#L52)).

So basically putting it below the systemjs script is a bit worse for perf, but always works. Putting it above the systemjs script is better for perf, but doesn't work with extras.

**Comment by guybedford** [comment]
This should be fixed in the latest refactoring that uses a three-phase script check.
